### ISSUE TITLE
End to end testing - "comparative values"

### ISSUE THREAD

**Issue by SuyDesignz** [jplag/JPlag#548]
@jplag/studdev 
@jplag/maintainer

The following issue will refer to the issue ([endToEndTesting-Issue](https://github.com/jplag/JPlag/issues/193) #193) and should be about the comparison values for the end-to-end testing.
Suitable comparison values for endToEndTesting are to be evaluated as well as their persistence in the project.

First of all, I wanted to list the possible properties and explain which values are available for the tests at all.
The endToEnd tests are performed using the JPlag API:
``` java
JPlagOptions options = new JPlagOptions(submissionTestPaths, new ArrayList<String>(), LanguageOption.JAVA);  
JPlag jplag = new JPlag(options);  
JPlagResult result = jplag.run();  
//for the fine-granular testing strategy, the result object contains only one comparison.
for (JPlagComparison jPlagComparison : result.getComparisons()) {
     System.out.println("Comparison of the stored values and the current equality values");
     assertEquals(Float.parseFloat(resultFromXML), jPlagComparison.similarity());
}
```
In the code example above, the similarity value of the test is compared with the stored values.

## Access to the possible comparative values
The returned result object is [JPlagComparison](https://github.com/jplag/JPlag/blob/master/jplag/src/main/java/de/jplag/JPlagComparison.java) which can be used to access the values of the comparison.
```java
[...]
    private final Submission firstSubmission;
    private final Submission secondSubmission;

    private final List<Match> matches;

    public JPlagComparison(Submission firstSubmission, Submission secondSubmission) {
        this.firstSubmission = firstSubmission;
        this.secondSubmission = secondSubmission;
        matches = new ArrayList<>();
    }
[...]
```
This is built up from two [Submission-objects](https://github.com/jplag/JPlag/blob/master/jplag/src/main/java/de/jplag/Submission.java) and a List\<[Match](https://github.com/jplag/JPlag/blob/master/jplag/src/main/java/de/jplag/Match.java)\> 

## Possible comparative values
The similarity(float) is already used as a comparative value in the current tests. 
The following possible comparison values must be discussed:
1. Here, the values "maximalSimilarity()" and "minimalSimilarity()" are additionally available. 
2. In the Match object length(int), startOfFirst(int) and startOfSecond(int) can be accessed.
3. The compared submission objects contain other possible comparative values (for every submission)
3.1 length of the token list 
3.2 the token list contains the JavaTokens with type number, hash value, length of the token and the line number
3.3 the token list also contains a list of token hashes which consists of a key-value pair. 

The usefulness of the tokenList was already briefly discussed in a meeting, but I would like to continue this here and put it up for discussion again
If anyone has more suggestions that could possibly be compared, feel free to continue in the comments.

## Format of the persistence file
Currently the comparison objects are stored in a *.xml file. It was suggested to change this to another file format for the final version. For this, *.json was already mentioned, which I also consider to be a useful change. If anyone would like to suggest a different format we can discuss that in the comments as well. Otherwise I will change the current format and the corresponding parser to the *.json format. 

Edit:
I have already changed the file format to json in the current version. So this does not need to be discussed anymore unless someone suggests something else.

**Comment by tsaglam** [comment]
@jplag/maintainer  @jplag/studdev

**Comment by tsaglam** [comment]
**In my opinion, we should check for each comparison:**

- similarity (which is avg)
- max similarity
- min similarity
- number of matches

**We should not check:**
- Match object length(int), startOfFirst(int) and startOfSecond(int)
- tokens/token list details like type number, hash value, length of the token, and the line number

**Open for discussion:** 

- Number of tokens for each submissions

**Comment by tsaglam** [comment]
Also, note that you need a delta for each similarity check, as you want to accept slightly different changes. This delta should be a central constant, thus allowing you to change it in one place for all of your tests. Maybe we need different constants for different metrics? As a starting point, there is a delta value in one of the existing test classes (TestBase maybe?).

**Comment by JanWittler** [comment]
First of all, thanks for the good description of the issue.

I think the general approach to discuss is the granularity of the comparison. Obviously, all metrics are linked together, thus if e.g. the max similarity changes we know that either the token list length or the number / size of matches changed too.
From my perspective, there are two reasonable approaches. First, on the fine-grained level of matches but rather on the abstraction of file line numbers instead of on the tokens. This allows the testing approach to be resilient against changes to the language frontend.
Second, on the coarse-grained level of similarity (min / max). This simplifies the expected test output description significantly with the drawback of requiring adjustments when language frontends change. Also, it will probably be hard to justify the choice of the delta.

Honestly, I am not sure which approach is better. If the second one works fine, I would go with it as it is already partly implemented. It is also probably easier to maintain and extend.
So to sum up, for the initial approach I would select:
- min similarity
- max similarity
- ~avg similarity (as it is just `(min + max) / 2`)~
- ~number of matches~
- ~number of tokens~

**Comment by tsaglam** [comment]
> So to sum up, for the initial approach I would select:

I agree, but I would argue for checking the number of matches for each comparison. This is between fine-grained and coarse-grained but allows for detecting changes in how code blocks are matched that do not affect the similarity.

**Comment by SuyDesignz** [comment]
Thanks in advance for the feedback and your expertise.

After the changes in the Current [Draft PR](https://github.com/jplag/JPlag/pull/551), I will expand the following comparison values in the tests and in the json file and then give feedback if it has an impact on change detection:
1. min similarity
2. max similarity
3. number of matches

**Comment by SuyDesignz** [comment]
The similarity was removed because it can be calculated by max and min. 
``` json
 "result_minimal_similarity" : 100.0,
 "result_maximum_similarity" : 100.0,
 "result_matched_token_number" : 56,
```
`- Issue closed -`
